src/dotnet/commands/dotnet-sln/add/Program.cs

```diff
@@ -16,12 +17,13 @@ internal class AddProjectToSolutionCommand : CommandBase
     {
         private readonly AppliedOption _appliedCommand;
         private readonly string _fileOrDirectory;
-        private readonly bool _inRoot;
-        
+        private readonly Func<SlnFile, string, IList<string>> _determineSolutionFolder;
```

Comment:
Rather than storing a `Func`, could we store the `IList<string>` of the solution folders?  I think it would be a little less confusing.

Reply:
It might make a mess of the code.[ In one of the three cases](https://github.com/dotnet/cli/blob/e84a3b763618d5c15645d76d574f835f4c5758a7/src/dotnet/commands/dotnet-sln/add/Program.cs#L55-L61), the solution folders list depends on the concrete project. The Func / delegate approach was the most concise approach I could come up with.

Should I move the if-logic to the loop over the projects?

Reply:
Personally, I think it would be easier to reason with a method that does the if-logic rather than the `Func`.

If we were storing `_inRoot` and `_relativeRoot` and moved the extension method logic into this class, then the method would simply be something like this:

```c#
private IList<string> GetSolutionFolders(SlnFile slnFile, string fullProjectPath)
{
    if (_inRoot)
    {
        return null;
    }

    if (!string.IsNullOrEmpty(_relativeRoot))
    {
        return _relativeRoot.Split(Path.DirectorySeparatorChar);
    }

    return GetSolutionFoldersFromProjectPath(
        Path.GetRelativePath(
            PathUtility.EnsureTrailingSlash(slnFile.BaseDirectory),
            fullProjectPath));
}
```

Reply:
@LordMike any thoughts to this remaining comment?  I see you moved the previous extension method into this class, so the above should be usable as is.

Reply:
Woops. Hadn't seen the update here.

I could do that, but would you want to recalculate the root in the `_relativeRoot` case, for each project to add?.. Seems wasteful, even though it's likely a matter of at most 1-3 times in a process lifetime (dotnet sln add ..)

Reply:
Sure, we can split it just once if it was specified and store that as a field instead of the string.